### PR TITLE
[xpu][fix] fix t5 sdpa pattern matcher

### PR DESCRIPTION
**Pull request by jianyizh**
spda pattern matcher does not working on current t5 models. I changed the mask to half to make it work again. 
TODO: 1. t5 training still does not match. 2. CUDA somehow enters the sdpa math path, so I disable it on CUDA. https://github.com/huggingface/transformers/blob/5f4ecf2d9f867a1255131d2461d75793c0cf1db2/src/transformers/models/t5/modeling_t5.py

Currently the backward graph of the model have some extra nodes that cannot be matched by serialized pattern. Maybe we missed some pass before joint graph
i.e. After set bias requires_grad=True,
```python
....
neg = neg(div)
fma = fma(neg, sum, mul)
convert = convert_element_type(fma)
view = view(convert, [32, 1024, 1024])
new_empty_strided = new_empty_strided(view); new_empty_strided = None # dead code but not eliminated before pattern match
view_1 = view(view, [4, 8, 1024, 1024]) # not in serialized pattern
view_2 = view(view_1, [32, 1024, 1024]) # not in serialized pattern
sum = sum(view_1, 0) # should be sum(fma)
bmm = bmm(xxx, view_2) # should be bmm(xxx, view)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 1]
```diff
@@ -581,54 +581,66 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
...
     query = query.permute(0, 2, 1, 3)
     key = key.permute(0, 2, 1, 3)
     value = value.permute(0, 2, 1, 3)
+    if attn_mask.device.type == "xpu":
+        attn_mask = attn_mask.contiguous()
     return _scaled_dot_product_attention(
         query,
         key,
         value,
-        attn_mask=attn_mask.to(dtype=query.dtype),
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/fuse_attention.py:605]: why does the new change not require converting to the `dtype` of the `query` tensor?

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 2]
```diff
@@ -581,54 +581,66 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
...
 
 
-def _sfdp_replacement_21(query, key, value, attn_mask):
+def _sfdp_replacement_21(query, key, value, attn_mask, dropout_p):
     counters["inductor"]["fuse_attention"] += 1
     query = query.permute(0, 2, 1, 3)
     key = key.permute(0, 2, 1, 3)
     value = value.permute(0, 2, 1, 3)
+    if attn_mask.device.type == "xpu":
+        attn_mask = attn_mask.contiguous()
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/fuse_attention.py]: We should move the logic to the XPU sdpa implementation. cc @LuFinch

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 3]
```diff
@@ -581,54 +581,66 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
...
     query = query.permute([0, 2, 1, 3])
     key = key.permute([0, 2, 1, 3])
     value = value.permute([0, 2, 1, 3])
     score = torch.matmul(query, key.permute(0, 1, 3, 2))
     masked_score = score + attn_mask
-    score = masked_score.type_as(query)
-    return score.float().softmax(dim=-1).type_as(query).matmul(value)
+    return torch.nn.functional.dropout(
+        masked_score.float().softmax(dim=-1).type_as(query), dropout_p
+    ).matmul(value)
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/fuse_attention.py]: What's the behavior if `dropout_p` is `None`? Meanwhile, the newly added `dropout_p` has changed the semantics of the original `_sfdp_pattern_21`. Why is it not a new pattern?

**Review by EikanWang — COMMENTED** [review]
Please add test cases.

**Review by EikanWang — APPROVED** [review]

**Comment by jianyizh** [comment]
@desertfire Can you please review this pr?

**Review by desertfire — APPROVED** [review]
